src/main/java/org/jenkinsci/plugins/docker/workflow/ImageNameTokens.java

```diff
@@ -50,16 +53,27 @@
     @Whitelisted
     public ImageNameTokens(@Nonnull String name) {
         int tagIdx = name.lastIndexOf(':');
+        String fullName;
         if (tagIdx != -1) {
-            this.userAndRepo = name.substring(0, tagIdx);
+            fullName = name.substring(0, tagIdx);
             if (tagIdx < name.length() - 1) {
                 this.tag = name.substring(tagIdx + 1);
             } else {
                 this.tag = "latest";
             }
         } else {
-            this.userAndRepo = name;
+            fullName = name;
             this.tag = "latest";
         }
+        int hostIdx = fullName.indexOf('/');
+        String hostname = (hostIdx != -1) ? fullName.substring(0, hostIdx) : null;
+        if (hostname != null && (hostname.equals("localhost") ||
+                Pattern.compile("[.:]").matcher(hostname).find())) {
```

Comment:
This logic is more or less lifted from https://github.com/docker/docker/blob/master/reference/reference.go